<a href="https://colab.research.google.com/github/535amar/st20338869_DAS7000_PORT1/blob/main/st20338869_DAS7000_PORT1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**StudentID: st20338869**

**Dataset: Global Dataset of Cyber Incidents**

[link text](https://zenodo.org/records/14965395)

**Aim:**   : The European Repository of cyber incidents(EuRepoC) is the releasing head of the Global Dataset of cyber incidents in V1.3 as the dataset i used for my assessment.in the data set contain over 3416 cyber incidents which has been started between 01.01.2000 and 2024. the trends in cyber world is changing day by day. Big data and resting data are vulnerable, as cyber threats continue to evolve rapidly, driven by the growth of big data and increasingly interconnected digital systems, organisational data and assets have become more vulnerable to cyber attacks. Understanding historical cyber incident data, attack patterns, and targeted entities is therefore critical for improving cyber resilience.
The aim of this analyse is to handle and explore, clean and do deep EDA inorder to identify key trends, attack types, affected regions, and targeted enities over time. The study of this data gives big insights that give support organisations in risk awarenes, and in formed decision making and more effective pre cybersecurity strategies.

In [28]:
import pandas as pd
import numpy as np
import plotly.express as px
import ast

In [29]:
df = pd.read_csv('eurepoc_global_dataset_1_3.csv')

In [30]:
df.head()

incident_id                                               name  \
0         4163   Russian State-Sponsored Actors Linked to GRU ...   
1         4161  Unknown threat actors stole Microsoft Azure ac...   
2         4160  Unspecified US intelligence agencies stole tra...   
3         4159  Unspecified US intelligence agency stole trade...   
4         4158  Unknown Threat Actors breached the Attorney Ge...   

                                         description  start_date  \
0  On 19 December 2024, a cyber attack attributed...  19.12.2024   
1  Unit 42 researchers uncovered a phishing campa...  01.06.2024   
2  The Chinese National Internet Emergency Respon...  01.05.2023   
3  The Chinese National Internet Emergency Respon...  01.08.2024   
4  In March 2024, a hacker breached the computer ...  01.03.2024   

        end_date                                inclusion_criterion  \
0     19.12.2024  Attack conducted by nation state (generic “sta...   
1  Not available        Attack on critical infrastructure target(s)   
2  Not available  Attack conducted by nation state (generic “sta...   
3  Not available  Attack conducted by nation state (generic “sta...   
4  Not available  Attack on (inter alia) political target(s), no...   

   inclusion_criterion_subcode  \
0  Not available;Not available   
1                Not available   
2                Not available   
3  Not available;Not available   
4                Not available   

                                   source_disclosure  \
0  Incident disclosed by authorities of victim state   
1          Incident disclosed by IT-security company   
2  Incident disclosed by authorities of victim state   
3  Incident disclosed by authorities of victim state   
4                                      Not available   

                               incident_type  \
0           Disruption;Hijacking with Misuse   
1           Data theft;Hijacking with Misuse   
2           Data theft;Hijacking with Misuse   
3           Data theft;Hijacking with Misuse   
4  Data theft & Doxing;Hijacking with Misuse   

                                       receiver_name  ...  \
0                      Ministry of Justice (Ukraine)  ...   
1  Not available;Not available;Not available;Not ...  ...   
2                                      Not available  ...   
3                        Not available;Not available  ...   
4            Attorney General's Office of Nuevo León  ...   

  legal_response_subtype legal_response_responding_country  \
0          Not available                           Ukraine   
1          Not available                     Not available   
2          Not available                     Not available   
3          Not available                     Not available   
4          Not available                     Not available   

     legal_response_responding_actor attribution_legal_reference  \
0  Security Service of Ukraine (SBU)               Not available   
1                      Not available               Not available   
2                      Not available               Not available   
3                      Not available               Not available   
4                      Not available               Not available   

  attribution_legal_reference_subcode  \
0                       Not available   
1                       Not available   
2                       Not available   
3                       Not available   
4                       Not available   

                                  response_indicator     casualties  \
0  Countermeasures under international law justif...  Not available   
1  Unfriendly acts/retorsions justified (missing ...  Not available   
2  Countermeasures under international law justif...  Not available   
3  Countermeasures under international law justif...  Not available   
4  Unfriendly acts/retorsions justified (missing ...  Not available   

                                          source_url added_to_db  updated_at  
0  https://www.t-on

In [31]:
df.shape

(3414, 85)

In [32]:
df.columns

Index(['incident_id', 'name', 'description', 'start_date', 'end_date',
       'inclusion_criterion', 'inclusion_criterion_subcode',
       'source_disclosure', 'incident_type', 'receiver_name',
       'receiver_country', 'receiver_country_alpha_2_code', 'receiver_regions',
       'receiver_category', 'receiver_subcategory', 'initiator_name',
       'initiator_country', 'initiator_alpha_2', 'initiator_category',
       'initiator_subcategory', 'number_attributions', 'attribution_id',
       'attribution_date', 'attribution_type', 'attribution_basis',
       'attributing_actor', 'attributing_company', 'attributing_country',
       'settled_initiator', 'initiator_country.1', 'initiator_category.1',
       'attribution_source_url', 'cyber_conflict_issue',
       'offline_conflict_issue', 'offline_conflict_name',
       'offline_conflict_intensity', 'offline_conflict_intensity_subcode',
       'number_political_responses', 'political_response_date',
       'political_response_type', 'politi

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3414 entries, 0 to 3413
Data columns (total 85 columns):
 #   Column                                 Non-Null Count  Dtype 
---  ------                                 --------------  ----- 
 0   incident_id                            3414 non-null   int64 
 1   name                                   3414 non-null   object
 2   description                            3414 non-null   object
 3   start_date                             3414 non-null   object
 4   end_date                               3414 non-null   object
 5   inclusion_criterion                    3414 non-null   object
 6   inclusion_criterion_subcode            3414 non-null   object
 7   source_disclosure                      3414 non-null   object
 8   incident_type                          3414 non-null   object
 9   receiver_name                          3414 non-null   object
 10  receiver_country                       3414 non-null   object
 11  receiver_country_

In [34]:
df.isna().sum()

incident_id           0
name                  0
description           0
start_date            0
end_date              0
                     ..
response_indicator    0
casualties            0
source_url            0
added_to_db           0
updated_at            0
Length: 85, dtype: int64

In [55]:
df.duplicated().sum()


np.int64(0)

The dataset contain nonstandard representation of missing values so change to standard cell name to understand the pands to catch the values of missing values

## Data quality issues identified
- Non-standard missing values: "Not available" appears in multiple fields and must be converted to NaN for correct missing-value detection.
- High sparsity: several variables (e.g., casualties, economic impact, political/legal response fields) are largely missing, limiting certain analyses.
- Multi-valued categoricals: some fields contain combined categories (e.g., incident_type, receiver_category) which increases unique values and reduces readability without aggregation.


In [35]:
df = df.replace("Not available", np.nan)

/tmp/ipython-input-3372635397.py:1: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



In [12]:
df.isna().sum().sort_values(ascending=False).head(10)

casualties                             3414
evidence_for_sanctions_indicator       3408
user_interaction                       3380
legal_response_subtype                 3372
zero_days_subcode                      3328
economic_impact                        3320
attribution_legal_reference_subcode    3187
political_response_subtype             3170
political_response_type                3147
attribution_legal_reference            3143
dtype: int64

The datasets has been loaded and inspected to understand the presence of missing values and it's strecture and it's data types

In [56]:
df['start_date'] = pd.to_datetime(df['start_date'], dayfirst=True, errors='coerce')
df['end_date'] = pd.to_datetime(df['end_date'], dayfirst=True, errors='coerce')

df['year'] = df['start_date'].dt.year
df["year"] = df["year"].astype("Int64")



In [37]:
df[['start_date', 'year']].head()

start_date    year
0 2024-12-19  2024.0
1 2024-06-01  2024.0
2 2023-05-01  2023.0
3 2024-08-01  2024.0
4 2024-03-01  2024.0

**Cyber incidents over time**

In [38]:
yearly = (
    df.dropna(subset=['year'])
      .groupby('year')
      .size()
      .reset_index(name='count')
)

import plotly.express as px

px.line(
    yearly,
    x='year',
    y='count',
    title='Global Cyber Incidents Over Time'
)


This shows that a noticiable acceleration is happeded after 2010. indicating both genuine rise in cyber activity and improved detection, attribution and reporting over time.

**choropleth map of cyber incidents affected countries**

*Question: Where are cyber incidents most frequently reported:*

*answer:*

In [54]:
country_counts = (
    df["receiver_country"]
    .replace("Not available", np.nan)
    .dropna()
    .value_counts()
    .reset_index()
)

country_counts.columns = ["country", "count"]


px.choropleth(
    country_counts,
    locations="country",
    locationmode="country names",
    color="count",
    color_continuous_scale="Viridis",
    title="Global Distribution of Cyber Incidents by Affected Country"
)


The choropleth map illustrates the global distribution of reported cyber incidents.
A higher concentration of incidents is visible in a small number of countries, which may reflect differences in digital infrastructure, geopolitical exposure, monitoring capability, and reporting transparency rather than true incident prevalence alone.
This highlights potential reporting bias and should be considered when interpreting geographic patterns.


**Top affected countries**

In [57]:
top_countries = (
    df['receiver_country']
    .value_counts()
    .head(10)
    .reset_index()
)

top_countries.columns = ['country', 'count']

px.bar(
    top_countries,
    x='country',
    y='count',
    title='Top 10 Affected Countries by Cyber Incidents'
)


**Distribution of cyber incident types**

In [39]:
incident_types = (
    df['incident_type']
    .value_counts()
    .reset_index()
)

incident_types.columns = ['incident_type', 'count']

px.bar(
    incident_types,
    x='incident_type',
    y='count',
    title='Distribution of Cyber Incident Types'
)


Here the result indicate that the types of incidents mostly occured,cyber incidents most compunded with incidents like data exfiltraction, disruption theft, hijacking are are reflect the how complex events rather than isolated attack types.

**Top 10 incidents**

In [40]:
top_incidents = (
    df['incident_type']
    .value_counts()
    .head(10)
    .reset_index()
)

top_incidents.columns = ['incident_type', 'count']

px.bar(
    top_incidents,
    x='incident_type',
    y='count',
    title='Top 10 Cyber Incident Types',
    labels={'incident_type': 'Incident Type', 'count': 'Number of Incidents'}
)


The incident type variable contains multiple combined categories for individual incidents, resulting in a large number of unique values.
To improve interpretability, the analysis focuses on the top 10 most frequently occurring incident type combinations.


**Top 10 targetd entity categories**

In [42]:
top_receivers = (
    df['receiver_category']
    .value_counts()
    .head(10)
    .reset_index()
)

top_receivers.columns = ['receiver_category', 'count']

px.bar(
    top_receivers,
    x='receiver_category',
    y='count',
    title='Top 10 Targeted Entity Categories',
    labels={'receiver_category': 'Targeted Entity Category', 'count': 'Number of Incidents'}
)


The targeted entity category variable includes multiple combined categories for individual incidents, resulting in a large number of unique values.
To improve clarity, the analysis focuses on the top 10 most frequently targeted entity categories.


In [58]:
tmp = (
    df["receiver_category"]
    .replace("Not available", np.nan)
    .dropna()
    .value_counts()
    .reset_index()
)
tmp.columns = ["receiver_category", "count"]

import plotly.express as px
px.treemap(
    tmp,
    path=["receiver_category"],
    values="count",
    title="Distribution of Cyber Incidents by Targeted Entity Category (Treemap)"
)


**Year and region heatmap**

the attack recivers region conatins multi valued lis like strings and some semicolon separated, so to analyse region trents it is normalised and exploded.

In [46]:
def parse_regions(x):
    if pd.isna(x):
        return []

    # Split on semicolon if multiple lists exist
    parts = str(x).split(";")
    regions = []

    for p in parts:
        p = p.strip()
        # Remove brackets
        p = p.strip("[]")
        # Split on commas
        items = p.split(",")
        for item in items:
            clean = item.replace("'", "").replace('"', "").strip()
            if clean and clean.lower() != "not available":
                regions.append(clean)

    return list(set(regions))  # removed duplicates


In [47]:
df_regions = df.copy()

df_regions["receiver_regions"] = df_regions["receiver_regions"].apply(parse_regions)

df_regions = df_regions.explode("receiver_regions")

df_regions = df_regions.dropna(subset=["receiver_regions", "year"])


In [48]:
heat = (
    df_regions
    .groupby(["year", "receiver_regions"])
    .size()
    .reset_index(name="count")
)


In [49]:
px.density_heatmap(
    heat,
    x="year",
    y="receiver_regions",
    z="count",
    title="Cyber Incidents by Year and Region",
    labels={"year": "Year", "receiver_regions": "Region", "count": "Number of Incidents"}
)


The receiver_regions variable contained complex, multi-valued entries, including multiple region lists stored as semicolon-separated strings.
A custom parsing function was implemented to normalise this field by extracting individual regions and exploding the data so that each incident–region relationship could be analysed separately.
This approach improves interpretability while acknowledging that a single cyber incident may affect multiple regions simultaneously.


**→ KEY OBSERVATION AND DATASET OVERVIEW:**

- The dataset contains aroud 3414 cyber incidents between 2000 and 2004.It includes 85 variables, covering temporal, geographic, political, legal, and technical attributes.
- Data types are mostly categorical with limited numerical fields.
- Missing values are presented as "Not available" not as NaN.So standardisation was required to allow pandas to correctly detect missing value. Missing values gives very high missingness it mostly impacted in legal and response related data. The dataset perfect for descriptive trend analyse, but weak for causal or econamic impact modelling.
- Feature engineering enabled trend analyse over time(year)
- Cyber incidents shows critical upward trend after 2010.
- Cyber incidents are distributed across all countries, but countries with lower counts may still experience attacks.
- The dataset also shows more cases are still there unreported.
- The most common incident types are: - Data theft
           - Disruption
           - Hijacking
- The dataset highlights more attacks are directly targets high value political and ecnomic targets.